nexosim/src/simulation.rs

```diff
@@ -739,6 +738,130 @@ pub struct DeadlockInfo {
     pub mailbox_size: usize,
 }
 
+/// An error returned upon failure during simulation state store procedure.
+#[non_exhaustive]
+#[derive(Debug)]
+pub enum SaveError {
+    #[cfg(feature = "server")]
+    /// Cbor serialization of the simulation's config has failed.
+    ConfigSerializationError(ciborium::ser::Error<std::io::Error>),
+    /// Failed attempt to binary encode model's state.
+    ModelEncodeError {
+        /// The fully qualified name of the model.
+        name: String,
+        /// Type name of the model.
+        type_name: &'static str,
+        /// `bincode` encoding error details.
+        encoding_error: bincode::error::EncodeError,
+    },
+    /// Failed attempt to binary encode an event.
+    EventEncodeError(usize, bincode::error::EncodeError),
+    /// Failed attempt to binary encode the scheduler queue.
+    SchedulerQueueEncodeError(bincode::error::EncodeError),
+    /// Failed attempt to binary encode the complete simulation state.
+    SimulationStateEncodeError(bincode::error::EncodeError),
+    /// Failed attempt to save an event with an unknown id.
+    EventNotFound(usize),
+    /// Argument data downcasting to a concrete type has failed.
+    ArgumentTypeMismatch(&'static str),
+    /// Failed attempt to binary encode an event argument.
+    ArgumentEncodeError(&'static str, bincode::error::EncodeError),
+}
```

Comment:
For consistency with `ModelEncodeError` I think I would rather see only struct-like variants such as:
```rust
EventEncodeError{
    source_id: usize,
    cause: bincode::error::EncodeError
}
```